src/gui/qgsrasterlayersaveasdialog.cpp

```diff
@@ -845,6 +907,41 @@ bool QgsRasterLayerSaveAsDialog::validate() const
   return true;
 }
 
+bool QgsRasterLayerSaveAsDialog::outputLayerExists() const
+{
+  QString uri;
+  if ( outputFormat() == QStringLiteral( "GPKG" ) )
+  {
+    uri = QStringLiteral( "GPKG:%1:%2" ).arg( outputFileName(), outputLayerName() );
+  }
+  else
+  {
+    uri = outputFileName();
+  }
+
+  std::unique_ptr< QgsRasterLayer > layer( new QgsRasterLayer( uri, "", QStringLiteral( "gdal" ) ) );
+  return layer->isValid();
+}
+
+void QgsRasterLayerSaveAsDialog::accept()
+{
+  if ( !validate() )
+  {
+    return;
+  }
+
+  if ( outputFormat() == QStringLiteral( "GPKG" ) && outputLayerExists() &&
+       QMessageBox::warning( this, tr( "Save Raster Layer" ),
+                             tr( "The layer %1 already exists in the target file, and overwriting layers in GeoPackage is not supported. "
```

Comment:
note: this is actually no longer true with GDAL 2.3.0 where GDALDatasetExecuteSQL(hDS, "DELLAYER:raster_name") should work. But that can be left for later